src/sagemaker/cli/compatibility/v2/modifiers/framework_version.py

```diff
@@ -112,30 +80,144 @@ def modify_node(self, node):
         - SKLearn: "0.20.0"
         - TensorFlow: "1.11.0"
 
+        The ``py_version`` value is determined by the framework, framework_version, and if it is a
+        model, whether the model accepts a py_version
+
         Args:
             node (ast.Call): a node that represents the constructor of a framework class.
         """
-        framework = self._framework_name_from_node(node)
-        node.keywords.append(
-            ast.keyword(arg="framework_version", value=ast.Str(s=FRAMEWORK_DEFAULTS[framework]))
-        )
+        framework, is_model = _framework_from_node(node)
 
-    def _framework_name_from_node(self, node):
-        """Retrieves the framework name based on the function call.
+        # if framework_version is not supplied, get default and append keyword
+        framework_version = _arg_value(node, FRAMEWORK_ARG)
+        if framework_version is None:
+            framework_version = FRAMEWORK_DEFAULTS[framework]
+            node.keywords.append(ast.keyword(arg=FRAMEWORK_ARG, value=ast.Str(s=framework_version)))
 
-        Args:
-            node (ast.Call): a node that represents the constructor of a framework class.
-                This can represent either <Framework> or sagemaker.<framework>.<Framework>.
+        # if py_version is not supplied, get a conditional default, and if not None, append keyword
+        py_version = _arg_value(node, PY_ARG)
+        if py_version is None:
+            py_version = _py_version_defaults(framework, framework_version, is_model)
+            if py_version:
+                node.keywords.append(ast.keyword(arg=PY_ARG, value=ast.Str(s=py_version)))
 
-        Returns:
-            str: the (capitalized) framework name.
-        """
-        if isinstance(node.func, ast.Name):
-            framework = node.func.id
-        elif isinstance(node.func, ast.Attribute):
-            framework = node.func.attr
 
-        if framework.endswith("Model"):
-            framework = framework[: framework.find("Model")]
+def _py_version_defaults(framework, framework_version, is_model=False):
+    """Gets the py_version required for the framework_version and if it's a model
+
+    Args:
+        framework (str): name of the framework
+
+    Returns:
+        str: the default py version, as appropriate. None if no default py_version
+    """
+    if framework in ["Chainer", "PyTorch"]:
+        return "py3"
+    if framework == "SKLearn" and not is_model:
+        return "py3"
+    if framework == "MXNet":
+        return "py2"
+    if framework == "TensorFlow" and not is_model:
+        return _tf_py_version_default(framework_version)
+    return None
+
+
+def _tf_py_version_default(framework_version):
+    """Gets the py_version default based on framework_version for TensorFlow."""
+    if not framework_version:
+        return "py2"
+    version = [int(s) for s in framework_version.split(".")]
+    if version < [1, 12]:
+        return "py2"
+    if version < [2, 2]:
+        return "py3"
```

Comment:
(optional) you could also use `packaging.version.Version` if you're getting tired of the split/list comprehension

Reply:
will update